### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by samodovdi**

### CODE REVIEW

File: src/Lykke.Service.GoogleAnalyticsWrapper.Core/Domain/TrackerEvents.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿namespace Lykke.Service.GoogleAnalyticsWrapper.Core.Domain
+{
+    public static class TrackerEvents
+    {
+        public const string UserRegistered = "UserRegistered";
```
Comment (starkmsu) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Core/Domain/TrackerEvents.cs]: Public constants require some description

File: src/Lykke.Service.GoogleAnalyticsWrapper.Core/Extensions/DistributedCacheExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        {
+            Locks = Enumerable
+                .Range(0, 1000)
+                .ToDictionary(
+                    i => i,
+                    i => new SemaphoreSlim(1, 1));
+
+        }
+
+        public static async Task<T> TryGetFromCacheAsync<T>(this IDistributedCache cache, string key, Func<Task<T>> getRecordFunc, TimeSpan? absoluteExpiration = null, TimeSpan? slidingExpiration = null)
```
Comment (starkmsu) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Core/Extensions/DistributedCacheExtensions.cs]: May be this should be renamed to GetOrAddAsync?

File: src/Lykke.Service.GoogleAnalyticsWrapper.Services/GaTrackerService.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            BaseAssetSettings baseAsset = null;
+            string assetType = null;
+
+            if (_trackAssetsSttings.Fiat.Assets.Contains(assetId))
+            {
+                assetType = "Fiat";
+                baseAsset = _trackAssetsSttings.Fiat.BaseAsset;
+            }
+
+            if (_trackAssetsSttings.Crypto.Assets.Contains(assetId))
```
Comment (starkmsu) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Services/GaTrackerService.cs:163]: missing 'else' ?

File: src/Lykke.Service.GoogleAnalyticsWrapper/Models/IsAliveResponse.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Collections.Generic;
+
+namespace Lykke.Service.GoogleAnalyticsWrapper.Models
+{
+    public class IsAliveResponse
```
Comment (starkmsu) [inline src/Lykke.Service.GoogleAnalyticsWrapper/Models/IsAliveResponse.cs]: Can be repalced with Lykke.Common.Api.Contract.Responses.IsAliveResponse

File: client/Lykke.Service.GoogleAnalyticsWrapper.Client/AutofacExtension.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+                .SingleInstance();
+        }
+
+        /// <summary>
+        /// Registers GoogleAnalyticsWrapperClient as IGoogleAnalyticsWrapperClient
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <param name="settings"></param>
+        public static void RegisterGoogleAnalyticsWrapperClient(this ContainerBuilder builder, GoogleAnalyticsWrapperServiceClientSettings settings)
+        {
```
Comment (MikhailGerasimov) [inline client/Lykke.Service.GoogleAnalyticsWrapper.Client/AutofacExtension.cs:36]: I would check here settings for null

File: client/Lykke.Service.GoogleAnalyticsWrapper.Client/Lykke.Service.GoogleAnalyticsWrapper.Client.xml [hunk 6]
```diff
@@ -0,0 +1,961 @@
+<?xml version="1.0"?>
```
Comment (MikhailGerasimov) [inline client/Lykke.Service.GoogleAnalyticsWrapper.Client/Lykke.Service.GoogleAnalyticsWrapper.Client.xml]: Exclude this file from Git. It is generated during build.

File: src/Lykke.Service.GoogleAnalyticsWrapper.AzureRepositories/GaTraffic/GaTrafficRepository.cs [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+        public GaTrafficRepository(INoSQLTableStorage<GaTrafficEntity> tableStorage)
+        {
+            _tableStorage = tableStorage;
+        }
+        
+        public async Task AddAsync(IGaTraffic model)
+        {
+            var entity = GaTrafficEntity.Create(model);
+            await _tableStorage.InsertOrMergeAsync(entity);
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper.AzureRepositories/GaTraffic/GaTrafficRepository.cs]: Awit is redundant here and below

File: src/Lykke.Service.GoogleAnalyticsWrapper.Core/Domain/Health/HealthIssuesCollection.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Lykke.Service.GoogleAnalyticsWrapper.Core.Domain.Health
+{
+    public class HealthIssuesCollection : IReadOnlyCollection<HealthIssue>
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Core/Domain/Health/HealthIssuesCollection.cs:6]: What for this class? Any List<> can be casted to IReadOnlyCollection

File: src/Lykke.Service.GoogleAnalyticsWrapper.Core/Extensions/DeviceInfoExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+            return GetCleanValue(result);
+        }
+
+        private static string GetCleanValue(string value)
+        {
+            return value?.Replace("<", string.Empty).Replace(">", string.Empty);
+        }
+
+        private static IDictionary<string, string> ParseUserAgent(string userAgent)
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Core/Extensions/DeviceInfoExtensions.cs]: Can be IReadOnlyDictionary

File: src/Lykke.Service.GoogleAnalyticsWrapper.Core/Extensions/DistributedCacheExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+using System.Threading.Tasks;
+using MessagePack;
+using MessagePack.Resolvers;
+using Microsoft.Extensions.Caching.Distributed;
+
+namespace Lykke.Service.GoogleAnalyticsWrapper.Core.Extensions
+{
+    public static class DistributedCacheExtensions
+    {
+        private static readonly IReadOnlyDictionary<int, SemaphoreSlim> Locks;
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Core/Extensions/DistributedCacheExtensions.cs:15]: Is this correct use *Local* locks with *Distributed* cache?

File: src/Lykke.Service.GoogleAnalyticsWrapper.Core/Settings/ServiceSettings/AssetListSettings.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Lykke.Service.GoogleAnalyticsWrapper.Core.Settings.ServiceSettings
+{
+    public class AssetListSettings
+    {
+        public string[] Assets { get; set; }
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Core/Settings/ServiceSettings/AssetListSettings.cs]: Can be IReadOnlyCollection

File: src/Lykke.Service.GoogleAnalyticsWrapper.Core/Settings/SlackNotifications/AzureQueuePublicationSettings.cs [hunk 12]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Lykke.Service.GoogleAnalyticsWrapper.Core.Settings.SlackNotifications
+{
+    public class AzureQueuePublicationSettings
+    {
+        public string ConnectionString { get; set; }
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Core/Settings/SlackNotifications/AzureQueuePublicationSettings.cs]: Add the check attribute

File: src/Lykke.Service.GoogleAnalyticsWrapper.Services/GaTrackerService.cs [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+                assetType = "Crypto";
+                baseAsset = _trackAssetsSttings.Crypto.BaseAsset;
+            }
+
+            return !string.IsNullOrEmpty(assetType) 
+                ? $"{(amount > 0 ? "Deposit" : "Withdraw")}{assetType}{(baseAsset.Multiplier == 1 ? string.Empty : $"_{multiplier:g2}")}_{baseAsset.AssetId}" 
+                : null;
+        }
+        
+        private async Task FillGaHitAsync<T>(T model) where T: GaBaseHit
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Services/GaTrackerService.cs]: I thing model type could be simple GaBaseHit. And below the same

File: src/Lykke.Service.GoogleAnalyticsWrapper.Services/GaUserService.cs [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+            _cache = cache;
+            _trackerUserRepository = trackerUserRepository;
+            _trafficRepository = trafficRepository;
+        }
+        
+        public async Task<string> GetGaUserIdAsync(string clientId)
+        {
+            var cachedValue = await _cache.TryGetFromCacheAsync(
+                GetUserIdKey(clientId),
+                async () =>
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Services/GaUserService.cs]: I would extract a method with a descriptive name

File: src/Lykke.Service.GoogleAnalyticsWrapper.Services/ShutdownManager.cs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    // If this is your case, use this class to manage shutdown.
+    // For example, sometimes some state should be saved only after all incoming message processing and 
+    // all periodical handler was stopped, and so on.
+    
+    public class ShutdownManager : IShutdownManager
+    {
+        private readonly ILog _log;
+        private readonly List<IStopable> _items = new List<IStopable>();
+
+        public ShutdownManager(ILog log)
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Services/ShutdownManager.cs:19]: Autofac can by self fill a collection of IStopable.
public ShutdownManager(ILog log, IEnumerable<IStopable>) should work

File: src/Lykke.Service.GoogleAnalyticsWrapper.Services/StartupManager.cs [hunk 16]
```diff
@@ -0,0 +1,30 @@
+﻿using System.Threading.Tasks;
+using Common.Log;
+using Lykke.Service.GoogleAnalyticsWrapper.Core.Services;
+
+namespace Lykke.Service.GoogleAnalyticsWrapper.Services
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper.Services/StartupManager.cs:5]: Delete the useless class

File: src/Lykke.Service.GoogleAnalyticsWrapper/Controllers/GoogleUserController.cs [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+using Lykke.Service.GoogleAnalyticsWrapper.Core.Domain.GaTraffic;
+using Lykke.Service.GoogleAnalyticsWrapper.Core.Services;
+using Lykke.Service.GoogleAnalyticsWrapper.Models;
+using Microsoft.AspNetCore.Mvc;
+using Swashbuckle.AspNetCore.SwaggerGen;
+
+namespace Lykke.Service.GoogleAnalyticsWrapper.Controllers
+{
+    [Route("api/[controller]")]
+    public class GoogleUserController : Controller
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper/Controllers/GoogleUserController.cs:15]: Add a description for public methods

File: src/Lykke.Service.GoogleAnalyticsWrapper/Controllers/TrackerController.cs [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+using Lykke.Service.GoogleAnalyticsWrapper.Core.Services;
+using Lykke.Service.GoogleAnalyticsWrapper.Models;
+using Microsoft.AspNetCore.Mvc;
+using Swashbuckle.AspNetCore.SwaggerGen;
+using ErrorResponse = Lykke.Common.Api.Contract.Responses.ErrorResponse;
+
+namespace Lykke.Service.GoogleAnalyticsWrapper.Controllers
+{
+    [Route("api/[controller]")]
+    public class TrackerController : Controller
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper/Controllers/TrackerController.cs:15]: Add a description for public methods

File: src/Lykke.Service.GoogleAnalyticsWrapper/Modules/ServiceModule.cs [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            builder.RegisterInstance(_log)
+                .As<ILog>()
+                .SingleInstance();
+
+            builder.RegisterType<HealthService>()
+                .As<IHealthService>()
+                .SingleInstance();
+
+            builder.RegisterType<StartupManager>()
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper/Modules/ServiceModule.cs:42]: Register as SingleInstance

File: src/Lykke.Service.GoogleAnalyticsWrapper/Modules/ServiceModule.cs [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+                .As<IStartupManager>();
+
+            builder.RegisterType<ShutdownManager>()
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.GoogleAnalyticsWrapper/Modules/ServiceModule.cs]: Register as SingleInstance

File: src/Lykke.Service.GoogleAnalyticsWrapper.Services/GaTrackerService.cs (hunk 3 above)
Comment (samodovdi) [reply src/Lykke.Service.GoogleAnalyticsWrapper.Services/GaTrackerService.cs:163]: not necessary

File: src/Lykke.Service.GoogleAnalyticsWrapper.Core/Domain/Health/HealthIssuesCollection.cs (hunk 8 above)
Comment (samodovdi) [reply src/Lykke.Service.GoogleAnalyticsWrapper.Core/Domain/Health/HealthIssuesCollection.cs:6]: its from the service template

**Review by MikhailGerasimov — APPROVED** [review]

**Pull request merged**